### ISSUE TITLE
Feature request: option to use SSH to forward port 2022

### ISSUE THREAD

**Issue by tbodt** [MisterTea/EternalTerminal#274]
In some environments, only connections on port 22 are allowed, not port 2022. It seems entirely possible to work around this by having Eternal Terminal spawn a command such as `ssh -L 2022:localhost:$RANDOM $host` and connect to the appropriate port on localhost instead of directly to the server. When the connection is lost, the port forwarding command can be restarted.

**Comment by MisterTea** [comment]
et doesn't persist an ssh connection, but only uses ssh at the beginning for handshaking and authentication (e.g. two-factor).

**Comment by tbodt** [comment]
That's the problem: most communication happens over a protocol other than SSH, which in my case isn't allowed by the firewall (and I'm in no position to change this). This issue is about forwarding all communication over SSH.

**Comment by MisterTea** [comment]
Yeah I understand your request but it's impossible. If we sent all the data over ssh, then we can't autoresume because ssh doesn't autoresume

**Comment by tbodt** [comment]
It should be possible to restart the ssh port forwarding command when it dies? Then treat this the same as a tcp connection disconnecting and reconnecting

**Comment by MisterTea** [comment]
@tbodt check out autossh. But if you are restarting the ssh connection, what's the point of using et?

**Comment by tbodt** [comment]
You still get the main benefit of et, which is that the shell doesn't get killed with the connection drops.

Can autossh work in combination with et? i.e. autossh forwards port 2022 and et talks over it?

**Comment by MisterTea** [comment]
Ok I see now, this is interesting.

Yeah I think you can forward port 2022 using ssh and then et to localhost to make the connection.  If this works, let's add it to the wiki

**Comment by gklanderman** [comment]
Hi @tbodt and @MisterTea, did either of you get this idea to work?  I just installed ET, and am attempting to do so, with no luck so far.

IIUC, I'll need to forward some port on the local host to the remote:22, so that ET can initially connect via ssh to start etserver?  And also forward local port 2022 to remote 2022, so that when local ET tries to connect to what it thinks is the remote port 2022 etserver, it actually gets there?

So I'm doing:

%  ssh -L localhost:3022:localhost:22 \                     
      -L localhost:2022:localhost:2022 \
      remote.host.com

At that point, I am able to ssh to the remote host via:

% ssh -o 'Port 3022' localhost

But then trying et fails:

% et --verbose 10 --logtostdout --ssh-option 'Port 3022' localhost
... lots of parsing ssh config ...
[INFO 2021-11-03 14:56:02,083 client-main TerminalClientMain.cpp:189] Parsed ssh config file, connecting to localhost
[V1 2021-11-03 14:56:02,083 client-main TerminalClientMain.cpp:14] Connecting
[V4 2021-11-03 14:56:02,084 client-main TcpSocketHandler.cpp:83] Before selecting sockFd
[V4 2021-11-03 14:56:02,084 client-main TcpSocketHandler.cpp:87] sockFd 6is selected6
[INFO 2021-11-03 14:56:02,084 client-main TcpSocketHandler.cpp:96] Connected to server: localhost using fd 6
[V1 2021-11-03 14:56:02,084 client-main UnixSocketHandler.cpp:166] Closing connection: 6
Error in authentication with etserver: Error:  Connection error communicating with et deamon: No such file or directory.

Any ideas what I'm doing wrong?

>  If this works, let's add it to the wiki

is there a wiki for ET?  The github wiki for the project does not seem to be active.

**Comment by MisterTea** [comment]
That error suggests that the et server is not running.  Can you telnet to localhost:2022 and verify that something is running on there?

**Comment by gklanderman** [comment]
Hmm maybe I'm misunderstanding, do I need to start etserver on the remote myself?  I thought that it would get started by et in the initial stage when it logs in via ssh.  Interestingly I see an error in my ssh session used to create the two port forwards when I run 'et',  before it has sshed in:

| channel 5: open failed: connect failed: Connection refused

I wouldn't expect it to connect to 2022 until after starting etserver, so I guess I am misunderstanding how this is supposed to work.

thank you for getting back so quickly!

**Comment by gklanderman** [comment]
I see etserver has a --bindip flag; what will it default to?  Can I override that from the et invocation (again assuming et starts etserver on the remote from the initial ssh login).  If the ssh port forward is to localhost:2022 on the remote host, and etserver is trying to bind all addresses on port 2022 that may be the problem.

**Comment by gklanderman** [comment]
hmm OK I see the ssh is invoking etterminal; etserver should already be running as a service.  I'll get back if I cannot get this figured out..

**Comment by gklanderman** [comment]
OK, so I got it working.. README.md doesn't include copying et.service into place in either of the sections on building ("Other Linux" -> Download and install from source", or "Building from Source" -> "Debian/Ubuntu") but I figured it out.

I edited /etc/et.cfg to have etserver listen only on localhost:

[Networking]
bind_ip = localhost

and it seems to be working as I expected in my first comment, thank you!

now to try X11 forwarding solution described in issue #291

**Comment by MisterTea** [comment]
Please edit the README to add anything that wasn't clear so people after you can use your experience

**Comment by gklanderman** [comment]
pull request #471 should address the README/build issue; it appears to just update the suggested build instructions to be: build the .deb and install it, rather than make install.  I presume installing the .deb will handle installing the service file etc.

**Comment by gklanderman** [comment]
So this workaround for ET not being able to use port 2022 directly is working, but in my case the ssh forwarding port 2022 cannot re-connect without my touching an authentication key.  I'm additionally using the technique from issue #291 to forward X11 but the problem is that I cannot turn off ClientAliveInterval in sshd config on the destination, so when the ssh forwarding port 2022 disconnects, and does not immediately reconnect (afk), the upper level ssh forwarding X11 over ET detects that the connection is down, and disconnects.  So I probably need to add real X11 forwarding to ET (or mosh) for this to really work for me.
